<?xml version="1.0"?>
<launch>
	<arg name="serial_no" default="" />
	<arg name="usb_port_id" default="" />
	<arg name="device_type" default="" />
	<arg name="json_file_path" default="" />
	<arg name="tf_prefix" default="camera" />
	<arg name="external_manager" default="false" />
	<arg name="manager" default="realsense2_camera_manager" />
	<arg name="base_frame_id" default="$(arg tf_prefix)_link" />
	<arg name="fisheye_width" default="640" />
	<arg name="fisheye_height" default="480" />
	<arg name="enable_fisheye" default="false" />
	<arg name="depth_width" default="640" />
	<arg name="depth_height" default="480" />
	<arg name="enable_depth" default="true" />
	<arg name="infra_width" default="640" />
	<arg name="infra_height" default="480" />
	<arg name="enable_infra1" default="false" />
	<arg name="enable_infra2" default="false" />
	<arg name="color_width" default="640" />
	<arg name="color_height" default="480" />
	<arg name="enable_color" default="true" />
	<arg name="fisheye_fps" default="30" />
	<arg name="depth_fps" default="15" />
	<arg name="infra_fps" default="30" />
	<arg name="color_fps" default="15" />
	<arg name="gyro_fps" default="400" />
	<arg name="accel_fps" default="250" />
	<arg name="enable_gyro" default="false" />
	<arg name="enable_accel" default="false" />
	<arg name="enable_pointcloud" default="true" />
	<arg name="pointcloud_texture_stream" default="RS2_STREAM_COLOR" />
	<arg name="pointcloud_texture_index" default="0" />
	<arg name="enable_sync" default="false" />
	<arg name="align_depth" default="false" />
	<arg name="publish_tf" default="true" />
	<arg name="tf_publish_rate" default="0" />
	<arg name="filters" default="" />
	<arg name="clip_distance" default="6.0" doc="important!! for CPU load decreasing" />
	<arg name="linear_accel_cov" default="0.01" />
	<arg name="initial_reset" default="false" />
	<arg name="unite_imu_method" default="" />
	<arg name="topic_odom_in" default="odom_in" />
	<arg name="calib_odom_file" default="" />
	<arg name="publish_odom_tf" default="false" />
	<arg name="allow_no_texture_points" default="false" />
	<!--
	launch realsense camera nodelet
	...
	-->
	<include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
		<arg name="tf_prefix" value="$(arg tf_prefix)" />
		<arg name="base_frame_id" value="$(arg base_frame_id)" />
		<arg name="external_manager" value="$(arg external_manager)" />
		<arg name="manager" value="$(arg manager)" />
		<arg name="serial_no" value="$(arg serial_no)" />
		<arg name="usb_port_id" value="$(arg usb_port_id)" />
		<arg name="device_type" value="$(arg device_type)" />
		<arg name="json_file_path" value="$(arg json_file_path)" />
		<arg name="enable_pointcloud" value="$(arg enable_pointcloud)" />
		<arg name="pointcloud_texture_stream" value="$(arg pointcloud_texture_stream)" />
		<arg name="pointcloud_texture_index" value="$(arg pointcloud_texture_index)" />
		<arg name="enable_sync" value="$(arg enable_sync)" />
		<arg name="align_depth" value="$(arg align_depth)" />
		<arg name="fisheye_width" value="$(arg fisheye_width)" />
		<arg name="fisheye_height" value="$(arg fisheye_height)" />
		<arg name="enable_fisheye" value="$(arg enable_fisheye)" />
		<arg name="depth_width" value="$(arg depth_width)" />
		<arg name="depth_height" value="$(arg depth_height)" />
		<arg name="enable_depth" value="$(arg enable_depth)" />
		<arg name="color_width" value="$(arg color_width)" />
		<arg name="color_height" value="$(arg color_height)" />
		<arg name="enable_color" value="$(arg enable_color)" />
		<arg name="infra_width" value="$(arg infra_width)" />
		<arg name="infra_height" value="$(arg infra_height)" />
		<arg name="enable_infra1" value="$(arg enable_infra1)" />
		<arg name="enable_infra2" value="$(arg enable_infra2)" />
		<arg name="fisheye_fps" value="$(arg fisheye_fps)" />
		<arg name="depth_fps" value="$(arg depth_fps)" />
		<arg name="infra_fps" value="$(arg infra_fps)" />
		<arg name="color_fps" value="$(arg color_fps)" />
		<arg name="gyro_fps" value="$(arg gyro_fps)" />
		<arg name="accel_fps" value="$(arg accel_fps)" />
		<arg name="enable_gyro" value="$(arg enable_gyro)" />
		<arg name="enable_accel" value="$(arg enable_accel)" />
		<arg name="publish_tf" value="$(arg publish_tf)" />
		<arg name="tf_publish_rate" value="$(arg tf_publish_rate)" />
		<arg name="filters" value="$(arg filters)" />
		<arg name="clip_distance" value="$(arg clip_distance)" />
		<arg name="linear_accel_cov" value="$(arg linear_accel_cov)" />
		<arg name="initial_reset" value="$(arg initial_reset)" />
		<arg name="unite_imu_method" value="$(arg unite_imu_method)" />
		<arg name="topic_odom_in" value="$(arg topic_odom_in)" />
		<arg name="calib_odom_file" value="$(arg calib_odom_file)" />
		<arg name="publish_odom_tf" value="$(arg publish_odom_tf)" />
		<arg name="allow_no_texture_points" value="$(arg allow_no_texture_points)" />
	</include>
	<!--  -->
	<!-- launch nodelet manager -->
	<!-- <node pkg="nodelet" type="nodelet" name="realsense_manager" args="manager">
		<arg name="num_worker_threads" default="4"/>
	</node> -->
	<!-- launch depth image to point cloud xzy -->
	<node name="depth_to_point_cloud" pkg="nodelet" type="nodelet" args="load depth_image_proc/point_cloud_xyz $(arg manager)">
		<remap from="camera_info" to="depth/camera_info" />
		<remap from="image_rect" to="depth/image_rect_raw" />
		<remap from="points" to="depth/points" />
	</node>
	<!-- filtering -->
	<node pkg="nodelet" type="nodelet" name="vg_filter" args="load pcl/VoxelGrid  $(arg manager)">
		<remap from="~input" to="depth/points" />
		<!-- <remap from="~output" to="/depth/filtered_points" /> -->
		<param name="input_frame" value="base_link" />
		<param name="filter_field_name" value="z" />
		<param name="filter_limit_min" value="-10.0" />
		<param name="filter_limit_max" value="10.0" />
		<param name="leaf_size" value="0.03" />
		<param name="filter_limit_negative" value="false" />
	</node>
	<node pkg="nodelet" type="nodelet" name="sor_filter" args="load pcl/StatisticalOutlierRemoval $(arg manager)">
		<remap from="~input" to="vg_filter/output" />
		<remap from="~output" to="depth/filtered_points" />
		<param name="input_frame" value="base_link" />
		<!-- <param name="filter_limit_min" value="-10.0" /> -->
		<!-- <param name="filter_limit_max" value="10.0" /> -->
		<param name="mean_k" value="10" />
		<param name="stddev" value="0.01" />
	</node>
</launch>
